### ISSUE TITLE
issues with advanced indexing of pytorch's tensor

### ISSUE THREAD

**Issue by OccupyMars2030** [OccupyMars2025/Protenix#14]
# The following tricks about the advanced indexing of pytorch's tensor are very enlightening for myself. So I leave this comment here for reference. These tricks are also related to this PR.
## run the following code  and pay attention to the following two lines. The function `gather_pair_embedding_in_dense_trunk` also comes from Protenix repo.
```python
    y = x[..., idx_q_expanded, idx_k_expanded, :]   # y.shape = [batch_dims, N_b, N_q, N_k, d]
    y = x[..., idx_q_expanded, :, :]   # y.shape =  [batch_dims, N_b, N_q, N_k, N_token, d]
```

```python
import torch

def gather_pair_embedding_in_dense_trunk(
    x: torch.Tensor, idx_q: torch.Tensor, idx_k: torch.Tensor
):
    """
    Selectively gather elements from a tensor using two sets of indices.
            N_b: number of blocks/trunks/splits
            N_q: number of queries
            N_k: number of keys

        x: [..., N_token, N_token, d]
        idx_q: [N_b, N_q]   
        idx_k: [N_b, N_k]

    Return:
        y: [..., N_b, N_q, N_k, d]
            where y[..., b, i, j, :] = x[..., idx_q[b, i], idx_k[b, j], :]
    """
    idx_q = idx_q.long()
    idx_k = idx_k.long()
    assert len(idx_q.shape) == len(idx_k.shape) == 2

    # Get the shape parameters
    N_b, N_q = idx_q.shape
    N_k = idx_k.shape[1]

    # Expand idx_q and idx_k to match the shape required for advanced indexing
    idx_q_expanded = idx_q.unsqueeze(-1).expand(-1, -1, N_k)  # shape: [N_b, N_q, N_k]
    idx_k_expanded = idx_k.unsqueeze(1).expand(-1, N_q, -1)   # shape: [N_b, N_q, N_k]

    # Use advanced indexing to gather the desired elements
    y = x[..., idx_q_expanded, idx_k_expanded, :]
    # y = x[..., idx_q_expanded, :, :]

    return y


# x: [..., N_token, N_token, d]
# idx_q: [N_b, N_q]   
# idx_k: [N_b, N_k]

batch_dims = 2
N_token = 10
d = 32
N_b, N_q, N_k = 4, 5, 6

x = torch.randn(batch_dims, N_token, N_token, d)
idx_q = torch.randint(0, N_token, (N_b, N_q))
idx_k = torch.randint(0, N_token, (N_b, N_k))

result = gather_pair_embedding_in_dense_trunk(x, idx_q, idx_k)
print(result.shape)  # Expected shape: [batch_dims, N_b, N_q, N_k, d]
```

## What I find can be just summarized in the following:
```python
import torch
import numpy as np

batch_dims = 2
N_token = 10
d = 32
N_b, N_q, N_k = 4, 5, 6

x = torch.randn(batch_dims, N_token, N_token, d)
idx_q = torch.randint(0, N_token, (N_b, N_q, N_k))
idx_k = torch.randint(0, N_token, (N_b, N_q, N_k))

result = x[:, idx_q, idx_k, :]
print("result.shape: ", result.shape)  # Expected shape: [batch_dims, N_b, N_q, N_k, d]
assert np.array_equal(result.numpy(), x.numpy()[:, idx_q.numpy(), idx_k.numpy(), :])

result_002 = x[:, idx_q, :, :]
print("result_002.shape: ", result_002.shape)  # Expected shape: [batch_dims, N_b, N_q, N_k, N_token, d]
assert np.array_equal(result_002.numpy(), x.numpy()[:, idx_q.numpy(), :, :])

result_003 = x[:, :, idx_k, :]
print("result_003.shape: ", result_003.shape)  # Expected shape: [batch_dims, N_token, N_b, N_q, N_k, d]
assert np.array_equal(result_003.numpy(), x.numpy()[:, :, idx_k.numpy(), :])

x = torch.randn(batch_dims, N_token, d, N_token)
idx_q = torch.randint(0, N_token, (N_b, N_q, N_k))
idx_k = torch.randint(0, N_token, (N_b, N_q, N_k))
result_004 = x[:, idx_q, :, idx_k]
print("result_004.shape: ", result_004.shape)  # This shape is very conterintuitive !!!
assert np.array_equal(result_004.numpy(), x.numpy()[:, idx_q.numpy(), :, idx_k.numpy()])

# output:
# result.shape:  torch.Size([2, 4, 5, 6, 32])
# result_002.shape:  torch.Size([2, 4, 5, 6, 10, 32])
# result_003.shape:  torch.Size([2, 10, 4, 5, 6, 32])
# result_004.shape:  torch.Size([4, 5, 6, 2, 32])

```

**Comment by OccupyMars2030** [comment]
# The following are explanation given by AI: 
# Explanation of `result_004.shape`

The shape `[4, 5, 6, 2, 32]` for `result_004` is due to how PyTorch handles **non-consecutive advanced indexing**.

## Key Concept

When using tensor indices (advanced indexing) in PyTorch:
- **Consecutive advanced indices** - Replace the original dimensions in place
- **Non-consecutive advanced indices** - Are grouped together at the **front** of the result shape

## Analysis of `result_004`

### Input shapes:
- `x.shape = (batch_dims, N_token, d, N_token) = (2, 10, 32, 10)`
- `idx_q.shape = (N_b, N_q, N_k) = (4, 5, 6)`
- `idx_k.shape = (N_b, N_q, N_k) = (4, 5, 6)`

### Indexing operation:
```python
result_004 = x[:, idx_q, :, idx_k]
```

### Breakdown by dimension:
```
x:        (2,    10,    32,    10)
Indexing: [:,  idx_q,   :,   idx_k]
           ↑    ↑       ↑     ↑
           |    |       |     |
        dim 0  dim 1  dim 2  dim 3
         (2)   (4,5,6) (32)  (4,5,6)
```

### Why the shape becomes `[4, 5, 6, 2, 32]`:

1. **Advanced indices (dim 1 and dim 3)** - These are **non-consecutive** (separated by a slice at dim 2), so PyTorch:
   - Broadcasts them together (they already have the same shape: 4,5,6)
   - Groups them at the **front** of the result shape → `[4, 5, 6, ...]`

2. **Non-advanced dimensions (dim 0 and dim 2)** - These are kept in their original order and appended after the advanced indices:
   - dim 0: size 2
   - dim 2: size 32
   - Result: `[..., 2, 32]`

### Final shape:
```
[4, 5, 6, 2, 32]
 ↑↑↑↑↑↑ ↑ ↑
 ││││││ │ │
 ││││││ │ └── dim 2 (size 32)
 ││││││ └───── dim 0 (size 2)
 ││││└─────── advanced indices (broadcasted)
 └└└└──────── idx_q and idx_k (non-consecutive)
```

## Comparison with other examples:

| Example | Indexing | Advanced Indices | Result Shape | Behavior |
|---------|----------|------------------|--------------|----------|
| `result` | `[:, idx_q, idx_k, :]` | Consecutive (dim 1-2) | `[2, 4, 5, 6, 32]` | Replace dims 1-2 in place |
| `result_002` | `[:, idx_q, :, :]` | Single (dim 1) | `[2, 4, 5, 6, 10, 32]` | Replace dim 1 in place |
| `result_003` | `[:, :, idx_k, :]` | Single (dim 2) | `[2, 10, 4, 5, 6, 32]` | Replace dim 2 in place |
| `result_004` | `[:, idx_q, :, idx_k]` | Non-consecutive (dim 1,3) | `[4, 5, 6, 2, 32]` | **Grouped at front** |

## Visualization

```
Original x:       (2,   10,   32,   10)
                   ┌───┬────┬────┬────┐
                   │ 2 │ 10 │ 32 │ 10 │
                   └───┴────┴────┴────┘
                     ↑    ↑    ↑    ↑
                     |    |    |    |
Indexing:          :  idx_q  :  idx_k
                     |    |    |    |
                     ↓    ↓    ↓    ↓
                   ┌──────┬───┬────┐
Result_004:        │4,5,6 │ 2 │ 32 │
                   └──────┴───┴────┘
                   ↑↑↑↑↑↑ ↑ ↑
                   ││││││ │ │
                   ││││││ │ └── Non-advanced dim 2 (size 32)
                   ││││││ └───── Non-advanced dim 0 (size 2)
                   └└└└└──────── Advanced indices (grouped at front)
```

## Summary

The counterintuitive shape `[4, 5, 6, 2, 32]` occurs because:
1. `idx_q` and `idx_k` are advanced indices at **non-consecutive positions** (dim 1 and dim 3)
2. PyTorch groups all non-consecutive advanced indices at the **front** of the result
3. Non-advanced dimensions (dim 0 and dim 2) are appended in their original order

This is standard behavior in PyTorch's advanced indexing mechanism, designed to handle complex indexing patterns efficiently.

**Comment by OccupyMars2030** [comment]
## Perfectly answer your question: https://numpy.org/doc/stable/user/basics.indexing.html#combining-advanced-and-basic-indexing
